pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -284,7 +291,7 @@ func balanceDomains(
 			// In other words, PTS can perform suboptimally if some of its chosen pods don't fit on other nodes.
 			// This is because the chosen pods aren't sorted, but immovable pods still count as "evicted" toward the PTS algorithm.
 			// So, a better selection heuristic could improve performance.
-			if !node.PodFitsAnyOtherNode(getPodsAssignedToNode, aboveToEvict[k], nodes) {
+			if !node.PodFitsAnyOtherNode(getPodsAssignedToNode, aboveToEvict[k], topologyNodesMap[sortedDomains[i].pair.value]) {
```

Comment:
This limits the check to nodes with the same topology domain value. Which makes the solution less suboptimal. Example:
Lets have the following distribution of pods among domains:
- nodeA1{zone=A, pods capacity=1}: 1
- nodeA2{zone=A, pods capacity=4}: 4
- nodeB1{zone=B, pods capacity=2}: 2
- nodeB2{zone=B, pods capacity=2}: 1

After getting processed:
```
sortedDomains = [
	topology{
		pair={key=zone,value=B},
		pods=[6,7,8],
	},
	topology{
		pair={key=zone,value=A},
		pods=[1,2,3,4,5],
	},
]
topologyNodesMap = {
	"A": [nodeA1, nodeA2],
	"B": [nodeB1, nodeB2],
}
skew=2 (5-3)
```

To get skew=1 one can move a pod from nodeA2 to nodeB2 since nodeA1 does not have resources for 2 pods. With the change applied in this PR `node.PodFitsAnyOtherNode` will be false as with `j=1` `topologyNodesMap[sortedDomains[1].pair.value] = topologyNodesMap["A"] = [nodeA1, nodeA2]`.

Did you consider this case?

Reply:
ping

Reply:
@a7i apologies for the delay. I was about to share the comment yesterday thought it was too late in the night to provide a failing unit test.

Reply:
@damemi also when running the unit tests ^^, [namespaceTopologySpreadConstraints](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/strategies/topologyspreadconstraint.go#L114) collects 30 items even though there's only one distinct constraint. Worth adding a check if a constraint already exists in `namespaceTopologySpreadConstraints`.

Reply:
Thanks for the details, that makes sense.

Also, regarding the 30 items, it is due to LabelSelector being a pointer to a struct, where the comparison is on the memory address rather than the content of the selector. I can look into patching this up.
<img width="613" alt="namespaceTopologySpreadConstraints" src="https://user-images.githubusercontent.com/9448877/173157466-1caccb26-4083-45f4-9994-59e0f4c0932e.png">

Reply:
I plan to take care of this namespaceTopologySpreadConstraints issue after this discussion unfolds :) I will keep an eye on it

Reply:
fyi, the test case with 6 domains seems to be a separate issue (fails on current and this new code). There are scenarios where the `j` position has less than `j-1` so it seems to me that we need to re-sort at the end of each iteration to ensure that index `j` is the largest in size.

Reply:
What about instead of passing `topologyNodesMap[sortedDomains[i].pair.value]` to pass all `topologyNodesMap[sortedDomains[i].pair.value]` for `i=0` until `i=l`, where `l` is the last domain that has size < `idealAvg`?

> There are scenarios where the j position has less than j-1 so it seems to me that we need to re-sort at the end of each iteration to ensure that index j is the largest in size.

I wonder if it is related to:
```
// skew = actual difference between the domains
skew := float64(len(sortedDomains[j].pods) - len(sortedDomains[i].pods))

// if k and j are within the maxSkew of each other, move to next belowOrEqualAvg
if int32(skew) <= constraint.MaxSkew {
	i++
	continue
}
```

I'd expect the condition to rather be:
```
if len(sortedDomains[i].pods) >= idealAvg {
	i++
	continue
}
```

Checking the code forth and back I don't see a need to re-sort the domains.

Reply:
Thanks for the hint 🥇 , I was thinking something along those lines. Will report back soon

Reply:
> we need to re-sort at the end of each iteration to ensure that index j is the largest in size.

I see it now. In the 6 domains test case these are the movement in the domains:
- len(sortedDomains[0].pods)=1, len(sortedDomains[5].pods=9
- len(sortedDomains[0].pods)=4, len(sortedDomains[5].pods=6
- len(sortedDomains[1].pods)=2, len(sortedDomains[5].pods=6
- len(sortedDomains[1].pods)=3, len(sortedDomains[5].pods=5
- len(sortedDomains[1].pods)=3, len(sortedDomains[4].pods=8
- len(sortedDomains[1].pods)=5, len(sortedDomains[4].pods=6
- len(sortedDomains[2].pods)=4, len(sortedDomains[4].pods=6
- len(sortedDomains[3].pods)=6, len(sortedDomains[4].pods=6

The third line causes another eviction from the last domain even though 6 pods in the last domain leads to "correct" re-distribution of pods. The domains need to be re-sorted as you say so a domain with the highest number of pods is always the one giving pods.

Reply:
After sorting the domains after each cycle the final distribution of pods is calculated as: [5, 4, 4, 5, 6, 6]. Without the sorting bits it is `[4, 5, 4, 6, 6, 5]`. With the same number of evictions.

I can't find an example where the final distribution of pods is wrong while keeping the domains unsorted. Not sure if sorting the pods will actually make any difference.

Reply:
I noticed that I needed to sort both i and j

Reply:
Experimenting with three options:
- no sorting
- sorting only the higher domains (when j-th domain's size gets decreased, move the domain to the left until all domains to the right have higher number of pods)
- sorting all the domains (when j-th domain's size gets decreased, move the domain to the left until all domains to the right have higher number of pods; when the i-th domain size gets increased, move the domain to the right until all domains to the left have lower number of pods)

Final distribution of all the options with the number of iterations of the `for i < j` cycle and evictions:
- [4, 5, 4, 6, 6, 5], iterations=7, evictions=6
- [5, 4, 4, 5, 6, 6], iterations=8, evictions=6
- [4, 4, 4, 6, 6, 6], iterations=7, evictions=5

Sequences:
- no sorting
  ```
  [1, 2, 4, 6, 8, 9]
  #[4, 2, 4, 6, 8, 6], evicted=3
  [4, 2, 4, 6, 8, 6]
  [4, 2, 4, 6, 8, 6]
  #[4, 3, 4, 6, 8, 5], evicted=1
  [4, 3, 4, 6, 8, 5]
  #[4, 5, 4, 6, 6, 5], evicted=2
  [4, 5, 4, 6, 6, 5]
  [4, 5, 4, 6, 6, 5]
  [4, 5, 4, 6, 6, 5]
  iterations: 7, all_evictions: 6, movements: 3
  ```
- the second half getting sorted
  ```
  [1, 2, 4, 6, 8, 9]
  #[4, 2, 4, 6, 6, 8], evicted=3
  [4, 2, 4, 6, 6, 8]
  #[5, 2, 4, 6, 6, 7], evicted=1
  [5, 2, 4, 6, 6, 7]
  [5, 2, 4, 6, 6, 7]
  #[5, 4, 4, 5, 6, 6], evicted=2
  [5, 4, 4, 5, 6, 6]
  [5, 4, 4, 5, 6, 6]
  [5, 4, 4, 5, 6, 6]
  [5, 4, 4, 5, 6, 6]
  iterations: 8, all_evictions: 6, movements: 3
  ```
- all sorted
  ```
  [1, 2, 4, 6, 8, 9]
  #[2, 4, 4, 6, 6, 8], evicted=3
  [2, 4, 4, 6, 6, 8]
  #[4, 4, 4, 6, 6, 6], evicted=2
  [4, 4, 4, 6, 6, 6]
  [4, 4, 4, 6, 6, 6]
  [4, 4, 4, 6, 6, 6]
  [4, 4, 4, 6, 6, 6]
  [4, 4, 4, 6, 6, 6]
  iterations: 7, all_evictions: 5, movements: 2
  ```

Reply:
@damemi was sorting the pods after each [movement of pods](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/strategies/topologyspreadconstraint.go#L311-L312) something you considered/abandoned when designing the algorithm? At this point it looks like an interesting improvement to the algorithm. Would be great to generate random sets of domains and measure the number of iterations/movements/evictions of the algorithm to see if there's any larger improvement.

Reply:
Very nice, thank you. I was trying to find out how to track which pods get evicted. We definitely need that assertion as part of our unit tests.

Reply:
Would you PTAL again? @ingvagabund @damemi @knelasevero 

- Balance domains: Ensure only nodes with pods below average count are considered as candidates for fit
- Test: Ensure specific pods were evicted (I did not do the assertion for ALL of the existing tests but we can do so after the [movements of pods](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/strategies/topologyspreadconstraint.go#L311-L312) changes)

Reply:
It looks good to me so far. Very cool to now have specific pod evictions being checked. I want to go over it a bit more consciously and circle back here.

Reply:
We have been running this in production and seeing a lot of improvements. Any chance this can be a patch candidate for 0.24?

Reply:
Not to sound lazy, but due to the work involved with patch releases and the timing for the next release, non-critical fixes are generally not worth the effort. I'm not opposed to it if there is a lot of demand though.

Reply:
Understood and makes sense. This was really urgent for us but we recently decided to run this fork for now.
![eviction-loop-same-node](https://user-images.githubusercontent.com/9448877/176938497-678500dd-6d00-4668-9a0f-c9b33fc88441.png).